### PR TITLE
Support save and open experiments

### PR DESCRIPTION
**Pull request by SparkSnail**

### CODE REVIEW

File: docs/en_US/Tutorial/Nnictl.md [hunk 1]
```diff
@@ -531,6 +528,62 @@ Debug mode will disable version check function in Trialkeeper.
...
+  > save an expeirment
+
+  ```bash
+  nnictl experiment save [experiment_id] --saveCodeDir
+  ```
+
+* __nnictl experiment load__
+  * Description
+
+    Load a nni experiment.
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/Nnictl.md:561]: an nni

File: docs/en_US/Tutorial/Nnictl.md [hunk 2]
```diff
@@ -531,6 +528,62 @@ Debug mode will disable version check function in Trialkeeper.
...
+
+    ```bash
+    nnictl experiment load [OPTIONS]
+    ```
+
+  * Options
+
+  |Name, shorthand|Required|Default|Description|
+  |------|------|------ |------|
+  |--path, -p|  True| |the full file path of nni package|
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/Nnictl.md:573]: 'the path of a saved nni experiment'. why it should be a full path?

File: docs/en_US/Tutorial/Nnictl.md [hunk 3]
```diff
@@ -531,6 +528,62 @@ Debug mode will disable version check function in Trialkeeper.
...
+    nnictl experiment save [OPTIONS]
+    ```
+
+  * Options
+
+  |Name, shorthand|Required|Default|Description|
+  |------|------|------ |------|
+  |id|  True| |The id of the experiment you want to save|
+  |--path, -p|  False| |the folder path to store nni experiment data, default current working directory|
+  |--saveCodeDir, -s| | |save codeDir data of the experiment, default False|
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/Nnictl.md:548]: miss Required field? False

File: docs/en_US/Tutorial/Nnictl.md [hunk 4]
```diff
@@ -531,6 +528,62 @@ Debug mode will disable version check function in Trialkeeper.
...
+    ```bash
+    nnictl experiment load [OPTIONS]
+    ```
+
+  * Options
+
+  |Name, shorthand|Required|Default|Description|
+  |------|------|------ |------|
+  |--path, -p|  True| |the full file path of nni package|
+  |--codeDir, -c| True| |the path of codeDir|
```
Comment (QuanluZhang) [inline docs/en_US/Tutorial/Nnictl.md:574]: could add more description for `codeDir`. for example, "the path that you want to put the code of the saved experiment into". BTW, what if the saved experiment does not include code?

File: tools/nni_cmd/nnictl.py [hunk 5]
```diff
@@ -149,6 +141,28 @@ def parse_args():
     parser_trial_export.add_argument('--type', '-t', choices=['json', 'csv'], required=True, dest='type', help='target file type')
     parser_trial_export.add_argument('--filename', '-f', required=True, dest='path', help='target file path')
     parser_trial_export.set_defaults(func=export_trials_data)
+    #save experiment data
+    parser_save_data = parser_experiment_subparsers.add_parser('save', help='save experiment data')
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl.py:145]: -> save an NNI experiment

File: tools/nni_cmd/nnictl.py [hunk 6]
```diff
@@ -149,6 +141,28 @@ def parse_args():
...
+    parser_save_data.add_argument('id', nargs='?', help='the id of experiment')
+    parser_save_data.add_argument('--path', '-p', required=False, help='the folder path to store nni experiment data, \
+                                   default current working directory')
+    parser_save_data.add_argument('--saveCodeDir', '-s', action='store_true', default=False, help='save codeDir data \
+                                   of the experiment')
+    parser_save_data.set_defaults(func=save_experiment)
+    #open experiment data
+    parser_open_data = parser_experiment_subparsers.add_parser('load', help='load experiment data')
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl.py:153]: -> load an NNI experiment

File: tools/nni_cmd/nnictl.py [hunk 7]
```diff
@@ -149,6 +141,28 @@ def parse_args():
...
+    parser_open_data.add_argument('--path', '-p', required=True, help='the path of nni package file')
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl.py:154]: better to make it consistent with the doc

File: tools/nni_cmd/nnictl_utils.py [hunk 8]
```diff
@@ -736,3 +739,162 @@ def search_space_auto_gen(args):
...
+        print_error('please set experiment id.')
+        exit(1)
+    if args.id not in experiment_dict:
+        print_error('Cannot find experiment {0}.'.format(args.id))
+        exit(1)
+    print_normal('Saving...')
+    nni_config = Config(experiment_dict[args.id]['fileName'])
+    logDir = os.path.join(os.path.expanduser("~"), 'nni-experiments', args.id)
+    if nni_config.get_config('logDir'):
+        logDir = os.path.join(nni_config.get_config('logDir'), args.id)
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl_utils.py:757]: better to use a function to generate `logDir`, this function can be put in a file which store constants or meta-datas. 'nni-experiments' can be seen as a type of metadata

File: tools/nni_cmd/nnictl_utils.py [hunk 9]
```diff
@@ -736,3 +739,162 @@ def search_space_auto_gen(args):
...
+    # Step1. Copy logDir to temp folder
+    if not os.path.exists(logDir):
+        print_error('logDir: %s does not exist!' % logDir)
+        exit(1)
+    temp_experiment_dir = os.path.join(temp_root_dir, 'experiment')
+    shutil.copytree(logDir, temp_experiment_dir)
+
+    # Step2. Copy nnictl metadata to temp folder
+    temp_nnictl_dir = os.path.join(temp_root_dir, 'nnictl')
+    os.makedirs(temp_nnictl_dir, exist_ok=True)
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl_utils.py:769]: is it possible that the dir already existed?

File: tools/nni_cmd/nnictl_utils.py [hunk 10]
```diff
@@ -736,3 +739,162 @@ def search_space_auto_gen(args):
...
+    src_path = os.path.join(os.path.join(temp_root_dir, experiment_id))
+    dest_path = os.path.join(os.path.join(logDir, experiment_id))
+    if os.path.exists(dest_path):
+        shutil.rmtree(dest_path)
+    shutil.copytree(src_path, dest_path)
+
+    # Step4. Copy code dir
+    codeDir = os.path.expanduser(args.codeDir)
+    if not os.path.isabs(codeDir):
+        codeDir = os.path.join(os. getcwd(), codeDir)
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl_utils.py:870]: os. getcwd()?

Comment (QuanluZhang) [reply tools/nni_cmd/nnictl_utils.py:870]: additional whitespace after dot

File: docs/en_US/Tutorial/Nnictl.md (hunk 1 above)
Comment (SparkSnail) [reply docs/en_US/Tutorial/Nnictl.md]: fixed

File: docs/en_US/Tutorial/Nnictl.md (hunk 2 above)
Comment (SparkSnail) [reply docs/en_US/Tutorial/Nnictl.md]: changed to `path`, not necessary to be `full path`

File: docs/en_US/Tutorial/Nnictl.md (hunk 3 above)
Comment (SparkSnail) [reply docs/en_US/Tutorial/Nnictl.md]: fixed

File: docs/en_US/Tutorial/Nnictl.md (hunk 4 above)
Comment (SparkSnail) [reply docs/en_US/Tutorial/Nnictl.md]: updated, if the experiment package does not code files, NNI will do nothing, just use new codeDir to replace old one.

File: tools/nni_cmd/nnictl.py (hunk 5 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl.py]: updated

File: tools/nni_cmd/nnictl.py (hunk 6 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl.py]: updated

File: tools/nni_cmd/nnictl.py (hunk 7 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl.py]: updated

File: tools/nni_cmd/nnictl_utils.py (hunk 8 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py]: updated

File: tools/nni_cmd/nnictl_utils.py (hunk 9 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py:769]: the dir name is generate by `random()` function, will not exist logically, I also add logic to handle this scenario, if the dir exist, nni will regenerate a new dir until it is not existed originally.

File: tools/nni_cmd/nnictl_utils.py (hunk 10 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py]: If the path is not a full path, nnictl will use current folder as a root path to find files.updated to `os.getcwd()`

File: tools/nni_cmd/common_utils.py [hunk 11]
```diff
@@ -83,3 +86,13 @@ def check_tensorboard_version():
...
         exit(1)
 
+def generate_temp_dir():
+    '''generate a temp folder'''
+    def generate_folder_name():
+        return os.path.join(tempfile.gettempdir(), 'nni', ''.join(random.sample(string.ascii_letters + string.digits, 8)))
+    temp_dir = generate_folder_name()
+    while os.path.exists(temp_dir):
+        temp_dir = generate_folder_name()
+    os.makedirs(temp_dir, exist_ok=True)
```
Comment (QuanluZhang) [inline tools/nni_cmd/common_utils.py:96]: as `temp_dir` does not exist, so it is ok to remove `exist_ok=True`?

File: tools/nni_cmd/nnictl_utils.py [hunk 12]
```diff
@@ -736,3 +736,165 @@ def search_space_auto_gen(args):
...
+    nni_config = Config(experiment_metadata.get('fileName'))
+    nnictl_exp_config = nni_config.get_config('experimentConfig')
+    if args.logDir:
+        logDir = args.logDir
+        nnictl_exp_config['logDir'] = logDir
+    else:
+        if nnictl_exp_config.get('logDir'):
+            logDir = nnictl_exp_config['logDir']
+        else:
+            logDir = os.path.join(os.path.expanduser("~"), 'nni-experiments')
```
Comment (QuanluZhang) [inline tools/nni_cmd/nnictl_utils.py:859]: also use `NNICTL_HOME_DIR` here

File: tools/nni_cmd/common_utils.py (hunk 11 above)
Comment (SparkSnail) [reply tools/nni_cmd/common_utils.py]: sure, updated

File: tools/nni_cmd/nnictl_utils.py (hunk 12 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py]: updated

File: tools/nni_cmd/nnictl_utils.py [hunk 13]
```diff
@@ -736,3 +736,165 @@ def search_space_auto_gen(args):
...
+
+def load_experiment(args):
+    '''load experiment data'''
+    package_path = os.path.expanduser(args.path)
+    if not os.path.exists(args.path):
+        print_error('file path %s does not exist!' % args.path)
+        exit(1)
+    temp_root_dir = generate_temp_dir()
+    shutil.unpack_archive(package_path, temp_root_dir)
+    print_normal('Opening...')
```
Comment (chicm-ms) [inline tools/nni_cmd/nnictl_utils.py:804]: `Loading...`  is better?

Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py]: updated

File: tools/nni_cmd/nnictl_utils.py [hunk 14]
```diff
@@ -736,3 +736,165 @@ def search_space_auto_gen(args):
...
+    nni_config.set_config('experimentConfig', nnictl_exp_config)
+    experiment_config.add_experiment(experiment_id,
+                                     experiment_metadata.get('port'),
+                                     experiment_metadata.get('startTime'),
+                                     experiment_metadata.get('fileName'),
+                                     experiment_metadata.get('platform'),
+                                     experiment_metadata.get('experimentName'),
+                                     experiment_metadata.get('endTime'),
+                                     experiment_metadata.get('status'))
+    print_normal('Open experiment %s succsss!' % experiment_id)
```
Comment (chicm-ms) [inline tools/nni_cmd/nnictl_utils.py:897]: Open -> Load ?

File: tools/nni_cmd/nnictl_utils.py [hunk 15]
```diff
@@ -736,3 +736,165 @@ def search_space_auto_gen(args):
...
+        shutil.rmtree(dest_path)
+    shutil.copytree(src_path, dest_path)
+
+    # Step4. Copy code dir
+    codeDir = os.path.expanduser(args.codeDir)
+    if not os.path.isabs(codeDir):
+        codeDir = os.path.join(os.getcwd(), codeDir)
+        print_normal('Expand codeDir to %s' % codeDir)
+    nnictl_exp_config['trial']['codeDir'] = codeDir
+    archive_code_dir = os.path.join(temp_root_dir, 'code')
```
Comment (chicm-ms) [inline tools/nni_cmd/nnictl_utils.py:873]: why not copy entire 'code'  dir using shutil.copytree?

Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py:873]: `shutil.copytree` only support copy entire folder to another folder, does not support merging two folder contents. We need to copy the contents under `code` folder, and do not change target folder name.

File: tools/nni_cmd/nnictl_utils.py (hunk 14 above)
Comment (SparkSnail) [reply tools/nni_cmd/nnictl_utils.py]: fixed

**Pull request merged**